### PR TITLE
[FIX] hr: change user on employee Access Error

### PR DESCRIPTION
**Pull request by kawkb**
Issue

- Admin unable to change related user on employee profile due to restricted access to bank account.

Steps to reproduce

1- In the Employee app, go to the HR settings tab of an employee.
2- Remove the related user from this profile.
3- Set that removed user as the related user on a different employee profile and save (expect no error).
4- Attempt to revert back to the original user.
5- Encounter an access error.

resolution

- Added sudo() in the browse operation to prevent access errors.

opw-3578412

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr/tests/test_hr_employee.py [hunk 1]
```diff
@@ -209,3 +209,25 @@ def test_employee_update_work_contact_id(self):
...
         self.assertEqual(employee_B.work_email, 'new_email@example.com')
+
+    def test_change_user_on_employee(self):
+        partner = self.env['res.partner'].create({
+            'name' : 'Test User Partner'
+            })
+        bank_account = self.env['res.partner.bank'].create({
+            'acc_number' : '1234567',
+            'partner_id' : partner.id
+            })
```
Comment (nle-odoo) [inline addons/hr/tests/test_hr_employee.py:220]: the test doesn't fail for me without the fix, I think the account should be on another company than `self.res_users_hr_officer` user to be able to reproduce it

**Review by nle-odoo — APPROVED** [review]
seems ok to me

the issue can be reproduced for example: 

- in belgian company (added by l10n_be)
- with "laurie poiret" employee (added by l10n_be_hr_contract_salary)
- when you set a "related user" of another company, the bank account of laurie poiret will be set to the company of the user_id.partner_id.company_id
- when you set "laurie poiret" back,  we will try to modify the bank account to set it back to belgian company, but because of access rules we don't have any access to this bank account (because he set to a company we don't have access to)

functionally I'm not sure if we should do the sudo as done here or if it's odd that we set a related user of another company than the company of the current employee

note: if this fix is kept, you can remove the `.sudo()` from 4 lines below the fix (`bank_account.sudo().allow_out_payment = False`) since the sudo is already applied

**Review by thle-odoo — COMMENTED** [review]
Hello,

Adding `sudo` solves the problem, but we shouldn't need it as we are the administrator and we have access to companies (not always).

We get an "Access Error" because we can't fetch a record from the `res.partner.bank` model with the query that contains the following where clause:
```
0:
'("res_partner_bank"."id" in %s)'
1:
'(("res_partner_bank"."company_id" in %s) OR "res_partner_bank"."company_id" IS NULL )' --> rule id: res_partner_bank_rule
```

This means that the user who modifies the user on the employee must have selected the same company as the partner on the bank account (if this partner has a company).

The expected behaviour is that we (as administrator) can make changes to the employee (that we have access) without selecting the company linked to the bank account according to the partner of user's employee.

Changing the rule (e.g. by using `user.company_ids.ids` instead of `company_ids`) will not update the current clients and will not solve the case where the bank account partner does not belong to a company to which we have access.

This scenario arises because there is no domain on `user_id` field in `hr.employee` model.
This will be resolved in the master with task 3640076.

So I agree with adding `sudo` for stable versions.

Thanks for your work !

CC: @Geoffrey-GMF

**Review by thle-odoo — COMMENTED** [review]
@kawkb
Another comment :)

It is also possible to add a domain for this field in stable version (the domain will be applied during the `name_search` method to generate the list of records available for the selection).

The ideal solution would have been to add a `check_company=True` directly to the `Many2one` field (https://github.com/odoo/odoo/blob/17.0/odoo/fields.py#L2955-L2957).

Unfortunately, this domain will be overwritten by the one used in the view: `domain="[('share', '=', False)]"`.

We can therefore complete the domain in the view as follows: `domain="[('company_ids', 'in', company_id), ('share', '=', False)]"`.
`company_id` will be the value of the `company_id` field for the current employee and this will force a search for users based on their company_ids (the authorised companies).

Note: pass an integer to the right node of a domain with `in` operator [expression.py](https://github.com/odoo/odoo/blob/saas-16.4/odoo/osv/expression.py#L1228-L1256)

Note2: the customer must update the view

To sum up, we can correct the domain for stable version (@Geoffrey-GMF)

**Comment by kawkb** [comment]
@thle-odoo can you please take a look at this?

**Comment by sofiagvaladze** [comment]
Hello @kawkb,

I agree with @thle-odoo, we can add domain in the view, that change is okey for stable. Though as noted by him, in order the change to be applied, the customer must update the view.

And also that change will be consistent in what will be done in master.

**Comment by kawkb** [comment]
Hello @sofiagvaladze can you please r+ if you think there are no changes to add!

File: addons/hr/tests/test_hr_employee.py [hunk 2]
```diff
@@ -209,3 +209,36 @@ def test_employee_update_work_contact_id(self):
...
+            'partner_id' : test_user.partner_id.id,
+        })
+        test_employee = self.env['hr.employee'].create({
+            'name': 'Test User - employee',
+            'user_id': test_user.id,
+            'company_id': test_company.id,
+            'bank_account_id': bank_account.id,
+        })
+        # change user -> bank account change company
+        with Form(test_employee) as employee_form:
```
Comment (sofiagvaladze) [inline addons/hr/tests/test_hr_employee.py:240]: Do we need Form to test changing user on an employee? probably not

File: addons/hr/tests/test_hr_employee.py [hunk 3]
```diff
@@ -209,3 +209,36 @@ def test_employee_update_work_contact_id(self):
         employee_B.work_email = 'new_email@example.com'
         self.assertEqual(employee_A.work_email, 'employee_A@example.com')
         self.assertEqual(employee_B.work_email, 'new_email@example.com')
+
+    @users('admin')
+    def test_change_user_on_employee(self):
+        test_other_user = self.env['res.users'].create({
+            'name': 'Test Ohter User',
```
Comment (sofiagvaladze) [inline addons/hr/tests/test_hr_employee.py:216]: typo

**Review by sofiagvaladze — CHANGES_REQUESTED** [review]
Two minor comments, otherwise it looks good to me

File: addons/hr/tests/test_hr_employee.py (hunk 2 above)
Comment (nle-odoo) [reply addons/hr/tests/test_hr_employee.py:240]: this is testing the scenario where we changed user_id and an onchange write on work_contact_id

we could directly write on user_id, work_contact_id, this would just test more lower level

Comment (sofiagvaladze) [reply addons/hr/tests/test_hr_employee.py:240]: I missed that, thanks

**Comment by sofiagvaladze** [comment]
robodoo r+
